src/core/bonded_interactions/angle_cosine.hpp

```diff
@@ -31,166 +31,101 @@
+    auto sin_phi0 = iaparams->p.angle_cosine.sin_phi0;
+    auto cos_phi0 = iaparams->p.angle_cosine.cos_phi0;
     // potential dependent term [dU/dphi = K * sin(phi - phi0)]
     // trig identity: sin(a - b) = sin(a)cos(b) - cos(a)sin(b)
-    pot_dep = K * (sin_phi * cos_phi0 - cos_phi * sin_phi0);
-  }
-
-  fac = pot_dep / sin_phi;
-
-  for (j = 0; j < 3; j++) {
-    fj[j] =
-        vec31[j] / (vec21_magn * vec31_magn) - cos_phi * vec21[j] / vec21_sqr;
-    fk[j] =
-        vec21[j] / (vec21_magn * vec31_magn) - cos_phi * vec31[j] / vec31_sqr;
-  }
-
-  // note that F1 = -(F2 + F3)
-  for (j = 0; j < 3; j++) {
-    force1[j] = force1[j] - fac * (fj[j] + fk[j]);
-    force2[j] = force2[j] + fac * fj[j];
-    force3[j] = force3[j] + fac * fk[j];
-  }
+    auto pot_dep = K * (sin_phi * cos_phi0 - cos_phi * sin_phi0);
+    auto fac = pot_dep / sin_phi;
+    return fac;
+  };
+
+  calc_angle_generic_3body_forces(p_mid, p_left, p_right, forceFactor, force1,
+                                  force2, force3);
 }
 
-/** Computes the three body angle interaction energy.
-    @param p_mid        Pointer to first particle.
-    @param p_left        Pointer to second/middle particle.
-    @param p_right        Pointer to third particle.
-    @param iaparams  bond type number of the angle interaction.
-    @param _energy   return energy pointer.
-    @return 0.
-*/
-inline int angle_cosine_energy(Particle *p_mid, Particle *p_left,
-                               Particle *p_right,
-                               Bonded_ia_parameters *iaparams,
+/** Computes the three-body angle interaction energy.
+ *  @param[in]  p_mid     Second/middle particle.
+ *  @param[in]  p_left    First/left particle.
+ *  @param[in]  p_right   Third/right particle.
+ *  @param[in]  iaparams  Bonded parameters for the angle interaction.
+ *  @param[out] _energy   Energy.
+ *  @retval 0
+ */
+inline int angle_cosine_energy(Particle const *p_mid, Particle const *p_left,
+                               Particle const *p_right,
+                               Bonded_ia_parameters const *iaparams,
                                double *_energy) {
-  double cosine, vec1[3], vec2[3], d1i, d2i, dist2;
-  int j;
-
-  /* vector from p_mid to p_left */
-  get_mi_vector(vec1, p_mid->r.p, p_left->r.p);
-  dist2 = sqrlen(vec1);
-  d1i = 1.0 / sqrt(dist2);
-  for (j = 0; j < 3; j++)
-    vec1[j] *= d1i;
-  /* vector from p_right to p_mid */
-  get_mi_vector(vec2, p_right->r.p, p_mid->r.p);
-  dist2 = sqrlen(vec2);
-  d2i = 1.0 / sqrt(dist2);
-  for (j = 0; j < 3; j++)
-    vec2[j] *= d2i;
+  /* vector from p_left to p_mid */
+  auto vec1 = get_mi_vector(p_mid->r.p, p_left->r.p);
+  double d1i = 1.0 / vec1.norm();
+  vec1 *= d1i;
+  /* vector from p_mid to p_right */
+  auto vec2 = get_mi_vector(p_right->r.p, p_mid->r.p);
+  double d2i = 1.0 / vec2.norm();
+  vec2 *= d2i;
   /* scalar product of vec1 and vec2 */
-  cosine = scalar(vec1, vec2);
+  double cosine = scalar(vec1, vec2);
```

Comment:
You can directly do
```c++
auto const cosine = vec1.normalize() * vec2.normalize()
```
it might be worth while to put the cosine calculation into a utility function, potentially including the angle sanitization,
as this is reused here already and also useful in different places.

Reply:
Most callers need the inverse length of the vectors besides the cosine, obtained by the `.norm()` method. I didn't use the `.normalize()` method to avoid calling `.norm()` twice for each vector. The dependency to `utils.hpp` was removed.